crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -194,6 +199,51 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
         }
     }
 
+    /// Generate a constraint from the *type* of the second argument of an `isinstance` call.
+    ///
+    /// Example: for `isinstance(…, str)`, we would infer `Type::ClassLiteral(str)` from the
+    /// second argument, but we need to generate a `Type::Instance(str)` constraint that can
+    /// be used to narrow down the type of the first argument.
+    fn to_isinstance_constraint(&self, classinfo: &Type<'db>) -> Option<Type<'db>> {
+        match classinfo {
+            Type::ClassLiteral(class) => Some(Type::Instance(*class)),
+            Type::Tuple(tuple) => {
+                let mut builder = UnionBuilder::new(self.db);
+                for element in tuple.elements(self.db) {
+                    builder = builder.add(self.to_isinstance_constraint(element)?);
+                }
+                Some(builder.build())
+            }
+            _ => None,
+        }
+    }
+
+    fn add_expr_call(&mut self, expr_call: &ast::ExprCall, expression: Expression<'db>) {
+        let scope = self.scope();
+        let inference = infer_expression_types(self.db, expression);
+
+        let func_name = expr_call.func.as_name_expr();
+
+        if matches!(func_name, Some(name) if name.id() == "isinstance") {
+            if expr_call.arguments.args.len() == 2 {
```

Comment:
We might also want to check that there are no keyword arguments here (`expr_call.arguments.keywords`). Neither mypy nor pyright does any type narrowing for this (which will obviously fail at runtime; `isinstance()` has no `foo` argument):

```py
class A: ...
class B: ...

def f(x: object):
    if isinstance(x, A, foo="bar"):
        reveal_type(x)
```

Reply:
Done.